### PR TITLE
Insertion sort

### PR DESCRIPTION
**Pull request by sjschmidt44**

### CODE REVIEW

File: insertion_sort.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+"""This module contains the insertion sort method, which performs an
+in-place sort on a passed in list. Insertion sort has a best case timeit
+complexity of O(n) when list is nearly sorted, and a worst case of O(n2)
+when the incoming list is already reverse sorted. While this not always
+the most efficient algorithm, it is still popular when the data is nearly
+sorted (because it is adaptive) or when the problem size is small.
+See the excellent 'sortingalgorithms.com' for more information.
+
+"""
```
Comment (joelstanner) [inline insertion_sort.py]: Awesome to see module documentation here.

File: insertion_sort.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+def in_sort(un_list):
+    """Sort a list in place using insertion sort.
+
+    args:
+        un_list: the list to sort
+    """
+    if type(un_list) is not list:
+        raise TypeError("You must pass a valid list as argument. Do it.")
+
```
Comment (joelstanner) [inline insertion_sort.py]: YES

File: insertion_sort.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        'from __main__ import BEST_CASE, insertion_sort').timeit(1000)
+
+    worst = Timer(
+        'insertion_sort({})'.format(WORST_CASE),
+        'from __main__ import WORST_CASE, insertion_sort').timeit(1000)
+
+    print("""Best case represented as a list that is already sorted\n
+        Worst case represented as a list that is absolute reverse of sorted""")
+    print('Best Case: {}'.format(best))
+    print('Worst Case: {}'.format(worst))
```
Comment (joelstanner) [inline insertion_sort.py]: Assignment calls for multiple list lengths: "demonstrates the performance characteristics of this sort _over a variety of lengths of input_ in both the best and worst-case scenarios.

File: test_insertion_sort.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    shuffle(unsorted)
+    in_sort(unsorted)
+    actual = unsorted
+    assert expected == actual
+
+
+def test_insertion_sort_wrong_type():
+    with pytest.raises(TypeError):
+        in_sort('some string')
+
```
Comment (joelstanner) [inline test_insertion_sort.py]: To further cover your code with tests, think about edge cases like what happens with a list with 0 or 1 items.

File: insertion_sort.py (hunk 1 above)
Comment (jonathanstallings) [inline insertion_sort.py]: Thanks for the recommendation to add it earlier.

File: insertion_sort.py (hunk 3 above)
Comment (jonathanstallings) [inline insertion_sort.py]: Good point; we missed this. I'll put in on our list of things to add on refactor.

File: test_insertion_sort.py (hunk 4 above)
Comment (jonathanstallings) [inline test_insertion_sort.py]: Right. That makes sense. I was coming up short when thinking of edge cases, since we were probing performance in the main module already. Those are good ideas. Thanks, @joelstanner .
